# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: browser/actors/AboutNewTabChild.jsm
# Commit: f64c8c158e52
# Full Hash: f64c8c158e52c96705cf1d507b9b30fe31305bd8
# Author: Mike Conley <mconley@mozilla.com>
# Date: 2020-04-23 09:52:48
# Description:
#   Bug 1631611 - Use a JSWindowActorChild to load about:newtab scripts from the ScriptPreloader rather than an observer. r=pdahiya
#   
#   This shifts the responsibility for ensuring that an about:home, about:welcome
#   or about:newtab page is where we inject the scripts to the JSWindowActor system,
#   rather than using the observer service on document creation and checking the
# ==============================================================================

diff -r af6964d3761b -r f64c8c158e52 browser/actors/AboutNewTabChild.jsm
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/browser/actors/AboutNewTabChild.jsm	Wed Apr 22 21:33:12 2020 +0000
@@ -0,0 +1,77 @@
+/* vim: set ts=2 sw=2 sts=2 et tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+"use strict";
+
+var EXPORTED_SYMBOLS = ["AboutNewTabChild"];
+
+const { Services } = ChromeUtils.import("resource://gre/modules/Services.jsm");
+const { XPCOMUtils } = ChromeUtils.import(
+  "resource://gre/modules/XPCOMUtils.jsm"
+);
+const { AppConstants } = ChromeUtils.import(
+  "resource://gre/modules/AppConstants.jsm"
+);
+
+XPCOMUtils.defineLazyPreferenceGetter(
+  this,
+  "ACTIVITY_STREAM_DEBUG",
+  "browser.newtabpage.activity-stream.debug",
+  false
+);
+
+XPCOMUtils.defineLazyPreferenceGetter(
+  this,
+  "SEPARATE_ABOUT_WELCOME",
+  "browser.aboutwelcome.enabled",
+  false
+);
+
+class AboutNewTabChild extends JSWindowActorChild {
+  handleEvent(event) {
+    if (event.type == "DOMContentLoaded") {
+      // If the separate about:welcome page is enabled, we can skip all of this,
+      // since that mode doesn't load any of the Activity Stream bits.
+      if (
+        SEPARATE_ABOUT_WELCOME &&
+        this.contentWindow.location.pathname.includes("welcome")
+      ) {
+        return;
+      }
+
+      // In the event that the document that was loaded here was the cached
+      // about:home document, then there's nothing further to do - the page
+      // will load its scripts itself.
+      //
+      // Note that it's okay to waive the xray wrappers here since this actor
+      // is registered to only run in the privileged about content process from
+      // about:home, about:newtab and about:welcome.
+      if (ChromeUtils.waiveXrays(this.contentWindow).__FROM_STARTUP_CACHE__) {
+        return;
+      }
+
+      const debug = !AppConstants.RELEASE_OR_BETA && ACTIVITY_STREAM_DEBUG;
+      const debugString = debug ? "-dev" : "";
+
+      // This list must match any similar ones in render-activity-stream-html.js.
+      const scripts = [
+        "chrome://browser/content/contentSearchUI.js",
+        "chrome://browser/content/contentSearchHandoffUI.js",
+        "chrome://browser/content/contentTheme.js",
+        `resource://activity-stream/vendor/react${debugString}.js`,
+        `resource://activity-stream/vendor/react-dom${debugString}.js`,
+        "resource://activity-stream/vendor/prop-types.js",
+        "resource://activity-stream/vendor/react-transition-group.js",
+        "resource://activity-stream/vendor/redux.js",
+        "resource://activity-stream/vendor/react-redux.js",
+        "resource://activity-stream/data/content/activity-stream.bundle.js",
+        "resource://activity-stream/data/content/newtab-render.js",
+      ];
+
+      for (let script of scripts) {
+        Services.scriptloader.loadSubScript(script, this.contentWindow);
+      }
+    }
+  }
+}